File: src-tauri/src/schema/patient.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum PatientState {
    Init,

    // 平稳期
    StableMild,     // 轻度
    StableModerate, // 中度
    StableSerious,  // 重度

    // 进展期
    ProgressiveMild,
    ProgressiveModerate,
    ProgressiveSerious,

    // 成熟期
    MatureMild,
    MatureModerate,
    MatureSerious,
}

impl PatientState {
    pub fn from_clinical(risser: i8, cobb: i32) -> Self {
        if cobb < 10 || cobb > 40 {
            return PatientState::Init;
        }

        if cobb < 20 {
            return match risser {
                0 => PatientState::StableMild,
                1 | 2 | 3 => PatientState::ProgressiveMild,
                _ => PatientState::MatureMild,
            };
        }

        if cobb < 40 {
            return match risser {
                0 => PatientState::StableModerate,
                1 | 2 | 3 => PatientState::ProgressiveModerate,
                _ => PatientState::MatureModerate,
            };
        }

        match risser {
            0 => PatientState::StableSerious,
            1 | 2 | 3 => PatientState::ProgressiveSerious,
            _ => PatientState::MatureSerious,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Patient {
    pub id: i64,
    pub registration_number: String,
    pub name: String,
    pub gender: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub menarche: Option<i8>,
    pub birthday: i64,
    pub school: String,
    pub grade: String,
    pub weight: f32,
    pub height: f32,
    pub contact: String,

    #[serde(default)]
    pub created_at: i64,

    pub state: PatientState,
}

File: src/lib/types/patient.ts
import { NonEmptyStringSchema } from "@/lib/schema";
import { z } from "zod";

export const PatientStateOptions = [
  "Init",
  "StableMild",
  "StableModerate",
  "StableSerious",
  "ProgressiveMild",
  "ProgressiveModerate",
  "ProgressiveSerious",
  "MatureMild",
  "MatureModerate",
  "MatureSerious",
];

export const GenderOptions = ["男生", "女生"];
export const GenderSchema = z.enum(GenderOptions);

export const GradeOptions = [
  "一年级",
  "二年级",
  "三年级",
  "四年级",
  "五年级",
  "六年级",
  "初一",
  "初二",
  "初三",
];
export const GradeSchema = z.enum(GradeOptions);

export const MenarcheSchema = z
  .number()
  .int()
  .min(1, "请填写正确的年龄")
  .max(20, "请填写正确的年龄")
  .optional();

export const WeightSchema = z
  .number({ message: "必须是有效的数字!" })
  .min(1, "请填写正确的体重")
  .max(300, "请填写正确的体重");

export const HeightSchema = z
  .number({ message: "必须是有效的数字!" })
  .min(1, "请填写正确的身高")
  .max(250, "请填写正确的身高");

export const PhoneNumberSchema = z.string().length(11, "请填写正确的手机号");

export const PatientSchema = z.object({
  id: z.number(),
  state: z.enum(PatientStateOptions),
  registration_number: z.string(),
  name: NonEmptyStringSchema,
  contact: NonEmptyStringSchema,
  height: HeightSchema,
  weight: WeightSchema,
  gender: GenderSchema,
  birthday: z.number(),
  school: NonEmptyStringSchema,
  grade: GradeSchema,
  menarche: MenarcheSchema,
  created_at: z.number(),
});

export type Patient = z.infer<typeof PatientSchema>;

export const DefaultPatient: Patient = {
  id: 0,
  state: "Init",
  registration_number: "",
  name: "",
  contact: "",
  height: 0,
  weight: 0,
  gender: "",
  birthday: 0,
  school: "",
  grade: "",
  created_at: 0,
};

export interface SearchParams {
  name?: string;
  registration_number?: string;
  contact?: string;
}
